File: Signature.js
// npm install js-sha1
// node Signature.js
const sha1 = require('js-sha1');
const PRIVATE_KEY = "PRIVATE_KEY";

class Signature {

    constructor() {
    }

    generate(params, isFromPc){
      let param = [];
      if(Object.keys(params).length){
         for(let i in params){
            param.push(params[i]);
         }
      }
      let salt = PRIVATE_KEY;
      param.push(salt);
      let timestamp = new Date().getTime();
      param.push(timestamp);
      let nonce = Math.ceil(Math.random() * 1000000);
      param.push(nonce);
      param.sort();
      param = param.join().replaceAll(',','');
      return sha1(param);
    }
}

// test case
let params = {
   "a" : "a",
   "b": "b"
}
let signObj = new Signature();
let signature = signObj.generate(params, true);
console.log(signature)

File: Signature.py
 # -*- coding:UTF-8 -*-
from pickle import TRUE
import time,random,hashlib

PRIVATE_KEY = "PRIVATE_KEY";

class Signature():
        
    def generate(self, params):  
        param = []
        for v in params.values():
            param.append(v)
        timestamp = int(round(time.time() * 1000))
        param.append(timestamp)
        nonce = random.randint(100000,999999)
        param.append(nonce)
        privateKey = PRIVATE_KEY
        param.append(privateKey)
        param.sort(key=lambda v: (isinstance(v, str), v))
        param = list(map(lambda x:str(x),param))
        paramStr = "".join(param)
        paramStr = hashlib.sha1(paramStr.encode("utf-8")).hexdigest()
        return paramStr;

# test case       
params = {
    "a" : "a",
    "b": "b"
}
signObj = Signature()
signature = signObj.generate(params)
print(signature) 